src/cli.c

```diff
@@ -28,153 +28,228 @@
 
+#include <stdlib.h>
 #include <string.h>
 
+#define EXIT_INVALID_ARG 2
 
-static int eval_buf(JSContext *ctx, const void *buf, int buf_len, const char *filename, int eval_flags) {
+#define OPT_PREFIX '-'
+#define OPT_ASSIGN '='
+
+#define is_longopt(opt, str) (opt.name && !strncmp(opt.name, str, opt.length))
+
+typedef struct CLIOption {
+    char key;
+    char *name;
+    unsigned length;
+} CLIOption;
+
+typedef struct Flags {
+    unsigned interactive, empty_run, strict_module_detection;
+    char *eval_expr, *override_filename;
+} Flags;
+
+static int eval_buf(JSContext *ctx, const char *buf, const char *filename, int eval_flags) {
     JSValue val;
-    int ret;
+    int ret = 0;
 
-    val = JS_Eval(ctx, buf, buf_len, filename, eval_flags);
+    val = JS_Eval(ctx, buf, strlen(buf), filename, eval_flags);
+    if (JS_IsException(val)) {
+        tjs_dump_error(ctx);
+        ret = -1;
+    }
+    JS_FreeValue(ctx, val);
+    return ret;
+}
+
+static int eval_module(JSContext *ctx, const char *filepath, char *override_filename, int eval_flags) {
+    JSValue val;
+    int ret = 0;
+
+    val = TJS_EvalFile(ctx, filepath, eval_flags, true, override_filename);
     if (JS_IsException(val)) {
         tjs_dump_error(ctx);
         ret = -1;
-    } else {
-        ret = 0;
     }
     JS_FreeValue(ctx, val);
     return ret;
 }
 
-void help(void) {
-    printf("tjs version %s\n"
-           "usage: tjs [options] [file]\n"
-           "-h  --help                       list options\n"
-           "-e  --eval EXPR                  evaluate EXPR\n"
-           "-i  --interactive                go to interactive mode\n"
-           "    --strict-module-detection    only run code as a module if its extension is \".mjs\"\n"
-           "    --override-filename FILENAME override filename in error messages\n"
-           "-q  --quit                       just instantiate the interpreter and quit\n",
-           tjs_version());
-    exit(1);
+static void print_help(void) {
+    printf("Usage: tjs [options] [file]\n"
+           "\n"
+           "Options:\n"
+           "  -v, --version                     print tjs version\n"
+           "  -h, --help                        list options\n"
+           "  -e, --eval EXPR                   evaluate EXPR\n"
+           "  -i, --interactive                 go to interactive mode\n"
+           "      --strict-module-detection     only run code as a module if its extension is \".mjs\"\n"
+           "      --override-filename FILENAME  override filename in error messages\n"
+           "  -q, --quit                        just instantiate the interpreter and quit\n");
+}
+
+static void print_version() {
+    printf("v%s\n", tjs_version());
+}
+
+static void report_bad_option(char *name) {
+    fprintf(stderr, "tjs: bad option -%s\n", name);
+}
+
+static void report_missing_argument(CLIOption *opt) {
+    if (opt->key)
+        fprintf(stderr, "tjs: -%c requires an argument\n", opt->key);
```

Comment:
Maybe we should have a define like PROG_NAME instead of repeting tjs all the time...